include/wil/cppwinrt_notifiable_module_lock.h

```diff
@@ -0,0 +1,96 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+//! @file
+//! Utilities for implementing OOP COM server with cppwinrt only
+
+#ifndef __WIL_CPPWINRT_NOTIFIABLE_MODULE_LOCK_INCLUDED
+#define __WIL_CPPWINRT_NOTIFIABLE_MODULE_LOCK_INCLUDED
+
+#ifdef WINRT_BASE_H
+#error You must include this header before including any winrt header
+#endif
+
+#ifndef WINRT_CUSTOM_MODULE_LOCK
+#error You must define 'WINRT_CUSTOM_MODULE_LOCK' at the project level
+#endif
+
+#include <atomic>
+#include <cstdint>
+
+namespace wil
+{
+// Adopted from cppwinrt
+struct notifiable_module_lock
+{
+    notifiable_module_lock() = default;
+
+    notifiable_module_lock(uint32_t count) : m_count(count)
+    {
+    }
+
+    uint32_t operator=(uint32_t count) noexcept
+    {
+        return m_count = count;
+    }
+
+    uint32_t operator++() noexcept
+    {
+        return m_count.fetch_add(1, std::memory_order_relaxed) + 1;
+    }
+
+    uint32_t operator--() noexcept
+    {
+        auto const remaining = m_count.fetch_sub(1, std::memory_order_release) - 1;
+
+        if (remaining == 0)
+        {
+            std::atomic_thread_fence(std::memory_order_acquire);
+            notifier();
+        }
+        else if (remaining < 0)
+        {
+            abort();
+        }
+
+        return remaining;
+    }
+
+    operator uint32_t() const noexcept
+    {
+        return m_count;
+    }
+
+    template <typename Func>
+    static void set_notifier(Func&& func)
+    {
+        notifier = std::forward<Func>(func);
+    }
+
+private:
+    std::atomic<int32_t> m_count{0};
+    static inline std::function<void()> notifier;

```

Comment:
Just noticed this; feels weird that this is a `static` variable, especially when `m_count` is not. I get that it's so that `set_notifier` can be a static member, but you can solve that problem in other ways. E.g. have a `static` `instance` method that `get_module_lock` also calls. E.g.
```c++
static notifiable_module_lock& instance() noexcept
{
    static notifiable_module_lock lock;
    return lock;
}
```
So then the usage becomes:
```c++
wil::notifiable_module_lock::instance().set_notifier(fn);
```
You could technically also have `set_notifier` remain `static`, but call `instance` to set the member.

Reply:
Right. I'm more inclined to make the whole class an explicit singleton. Right now it is a combination of singleton (the notifier) and non-singleton (the count). It wasn't a problem in cppwirnt even though there is supposed to be only 1 module lock because it was in the `impl` namespace and users are not supposed to directly use it

I feel like making the ctor private, and only provide the static `instance()` is the right way to go. But I'd have to figure out whether it still allows users to provide custom module lock composed of `wil::notifiable_module_lock` in some way.

Reply:
End up making ctor protected as I found customizing a singleton unpleasant. You basically have to implement the module lock again and forward all method calls to the singleton.

API wise I think singleton makes sense, but I'm not sure if it is acceptable by wil's standard. E.g. memory concern if the static instance is not used (although I bet if the static instance is not used, compiler is probably going to trim it).

Reply:
I think it feels a little awkward to have a singleton class with an `instance` method be designed such that users can derive from it. You could consider splitting it into two types: a "base type" that just has all the methods, data members, etc. that does not have an `instance` method, and an "impl" type that derives from this base type and exposes just an `instance` method.

> E.g. memory concern if the static instance is not used

Function statics are not initialized until the first time they are accessed. I'd also like to believe that the linker is smart enough to prune unused memory if the function is never called. That said, the variable will not exist unless this header is included, and this header wouldn't ever be included unless the consumer _wants_ that variable to exist.